api/tests/Behat/bootstrap/v2/Common/ExpectedResultsTrait.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Tests\Behat\v2\Common;
+
+trait ExpectedResultsTrait
+{
+    private array $comparisonItems = [];
+
+    public function expectedResultsDisplayed($sectionNumber, $expectedResultArrays, $context)
+    {
+        $xpath = '//dl|//tbody';
+        $topLevelTables = $this->getSession()->getPage()->findAll('xpath', $xpath);
+
+        $sections = [];
+        foreach ($topLevelTables as $tableItem) {
+            $this->comparisonItems = [];
+            if ('dl' == $tableItem->getTagName()) {
```

Comment:
This section may be more readable if we extract the different responsibilities out, something like:

```
        foreach ($topLevelTables as $tableItem) {
            $this->comparisonItems = [];
            
            $this->validateTableItem($tableItem);
            $this->handleDescriptionList($tableItem);
            $this->handleTableBody($tableItem);
            
            $sections[] = $this->comparisonItems;
        }
```

```
    private function validateTableItem(NodeElement $tableItem)
    {
        $htmlTag = $tableItem->getTagName();
        
        if (!in_array($htmlTag, ['tbody', 'dl'])) {
            $this->throwContextualException('The element being asserted on does not contain a dl or tbody element');
        }
    }
```

```
    private function handleDescriptionList(NodeElement $tableItem)
    {
        if ('dl' == $tableItem->getTagName()) {
            $xpath = '//li';
            $listItems = $tableItem->findAll('xpath', $xpath);
            
            if (count($listItems) > 0) {
                $this->getTextItems($listItems);
            } else {
                $xpath = '//dt|//dd';
                $descriptionDataItems = $tableItem->findAll('xpath', $xpath);
                $this->getTextItems($descriptionDataItems);
            }
        }
    }
```

```
    private function handleTableBody(NodeElement $tableItem)
    {
        if ('tbody' == $tableItem->getTagName()) {
            $xpath = '//tr';
            $tableRowItems = $tableItem->findAll('xpath', $xpath);
    
            if (count($tableRowItems) > 0) {
                foreach ($tableRowItems as $tableRowItem) {
                    $xpath = '//td|//th';
                    $tableDataItems = $tableRowItem->findAll('xpath', $xpath);
                    $this->getTextItems($tableDataItems);
                }
            } else {
                $xpath = '//td|//th';
                $tableDataItems = $tableItem->findAll('xpath', $xpath);
                $this->getTextItems($tableDataItems);
            }
        }
    }
```